servicetalk-loadbalancer-experimental/src/main/java/io/servicetalk/loadbalancer/RoundRobinSelector.java

```diff
@@ -78,6 +104,92 @@ protected Single<C> selectConnection0(
 
     @Override
     public HostSelector<ResolvedAddress, C> rebuildWithHosts(@Nonnull List<? extends Host<ResolvedAddress, C>> hosts) {
-        return new RoundRobinSelector<>(index, hosts, getTargetResource(), failOpen);
+        return new RoundRobinSelector<>(index, hosts, getTargetResource(), failOpen, ignoreWeights);
+    }
+
+    private static Scheduler buildScheduler(AtomicInteger index, List<? extends Host<?, ?>> hosts) {
+        boolean allEqualWeights = true;
+        double maxWeight = 0;
+        for (Host<?, ?> host : hosts) {
+            double hostWeight = host.weight();
+            maxWeight = Math.max(maxWeight, hostWeight);
+            allEqualWeights = allEqualWeights && approxEqual(hosts.get(0).weight(), hostWeight);
+        }
+
+        if (allEqualWeights) {
+            return new ConstantScheduler(index, hosts.size());
+        } else {
+            double scaleFactor = MAX_WEIGHT / maxWeight;
+            int[] scaledWeights = new int[hosts.size()];
+
+            for (int i = 0; i < scaledWeights.length; i++) {
+                // Using ceil ensures a few things:
+                // - our max weighted element is picked on every round
+                // - hosts with weights near zero will never be truly zero
+                // - true zero weights will still be truly zero
+                scaledWeights[i] = Math.min(MAX_WEIGHT, (int) Math.ceil(hosts.get(i).weight() * scaleFactor));
+            }
+            return new StrideScheduler(index, scaledWeights);
+        }
+    }
+
+    private abstract static class Scheduler {
+        abstract int nextHost();
+    }
+
+    private static final class ConstantScheduler extends Scheduler {
+
+        private final AtomicInteger index;
+        private final int hostsSize;
+
+        ConstantScheduler(AtomicInteger index, int hostsSize) {
+            this.index = index;
+            this.hostsSize = hostsSize;
+        }
+
+        @Override
+        int nextHost() {
+            return (int) (Integer.toUnsignedLong(index.getAndIncrement()) % hostsSize);
+        }
+    }
+
+    // The stride scheduler is heavily inspired by the Google gRPC implementations with some minor modifications.
+    // The stride scheduler algorithm is convenient in that  it doesn't require synchronization like a priority queue
+    // based solution would which fits well with highly concurrent nature of our HostSelector abstraction.
+    // See the java-grpc library for more details:
+    // https://github.com/grpc/grpc-java/blob/da619e2b/xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
+    private static final class StrideScheduler extends Scheduler {
+
+        private final AtomicInteger index;
+        private final int[] weights;
```

Comment:
If here we need to use `int`, wonder if it makes more sense to switch to `int` in our API instead of `double`. This is related to discussion opened by @daschl in another PR: https://github.com/apple/servicetalk/pull/2892#discussion_r1578922871

Reply:
I think it's fine: if we switch to int for weight we introduce the question of what is default and also have to change up the P2C weight selector which naturally operates on floating point types. At a minimum we should address it in a different commit.